power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -714,52 +766,174 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
+                    update_data.get_buffer_span_all_scenarios<meta_data::update_getter_s, CT>(), result);
+            }
+            if (comp_index >= 0 && result.uniform) {
+                result.elements_ps_in_update =
+                    update_data.get_component_info(comp_index).elements_per_scenario; // -1 for sparse
+            }
+            return result;
         };
 
-        // check all components
-        auto const update_independent = run_functor_with_all_types_return_array(is_component_update_independent);
-        return std::ranges::all_of(update_independent, [](bool const is_independent) { return is_independent; });
+        // check and return indenpendence of all components
+        return run_functor_with_all_types_return_vector(check_each_component);
+    }
+
+    std::vector<std::pair<std::string, bool>> is_update_independent(ConstDataset const& update_data) {
+        std::vector<std::pair<std::string, bool>> result;
+
+        // If the batch size is (0 or) 1, then the update data for this component is 'independent'
+        if (update_data.batch_size() <= 1) {
+            result.emplace_back("all component", true);
+            return result;
+        }
+
+        auto const all_comp_update_independence = check_components_independence(update_data);
+        for (auto const& comp : all_comp_update_independence) {
+            result.emplace_back(comp.name, comp.is_independent());
+        }
+
+        return result;
+    }
+    void validate_update_data_independence(UpdateCompProperties const& comp) const {
+        if (!comp.has_id && comp.ids_all_na) {
+            return; // empty dataset is still supported
+        }
+        if (comp.elements_in_base < comp.elements_ps_in_update) {
+            throw DatasetError("Update data has more elements per scenario than input data for component " + comp.name +
+                               "!");
+        }
+        if (comp.ids_part_na) {
+            throw DatasetError("Part of the IDs are not valid for component " + comp.name + " in update data!");
+        }
+        if (!comp.uniform) {
+            if (comp.is_columnar && !comp.has_id) {
+                throw DatasetError("Columnar input data without IDs for component " + comp.name + " is not uniform!");
+            }
+            if (!comp.is_columnar && comp.ids_all_na) {
+                throw DatasetError("Row based input data with all NA IDs for component " + comp.name +
+                                   " is not uniform!");
+            }
+        }
+        if (comp.elements_in_base != comp.elements_ps_in_update) {
+            if (comp.is_columnar && !comp.has_id) {
+                throw DatasetError("Columnar input data for component " + comp.name +
+                                   " has different number of elements per scenario in update and input data!");
+            }
+            if (!comp.is_columnar && comp.uniform && (comp.has_id && comp.ids_all_na)) {
```

Comment:
why is this logic so different? I know row-based id-free is not yet supported, but other than that, columnar and row-based should act exactly the same

in particular, for columnar, it could be possible to specify an `id` attribute column that contains all `nan`s. it doesn't make sense to do so, but I can think of cases where this could happen. columnar and row-based should act the same way; just sliced along a different axis

Reply:
Resolved in https://github.com/PowerGridModel/power-grid-model/pull/818